File: backend/drone-service/app/hardware/minio.py
import io
from datetime import datetime
from minio import Minio
from minio.error import S3Error
from config.config import settings


class MinIOClient:
    def __init__(self):
        self.client = Minio(
            settings.MINIO_ENDPOINT,
            access_key=settings.MINIO_ROOT_USER,
            secret_key=settings.MINIO_ROOT_PASSWORD,
            secure=settings.MINIO_USE_SSL
        )
        self.bucket_name = settings.MINIO_BUCKET_RECORDS
    
    async def upload_frame(self, drone_id: str, delivery_id: str, frame_data: bytes, frame_number: int) -> str | None:
        try:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            object_name = f"{drone_id}/{delivery_id}/{timestamp}_frame_{frame_number}.jpg"
            
            stream = io.BytesIO(frame_data)
            self.client.put_object(
                self.bucket_name,
                object_name,
                stream,
                length=len(frame_data),
                content_type="image/jpeg"
            )
            return object_name
        except S3Error as e:
            print(f"Error uploading frame: {e}")
            return None

File: backend/drone-service/pkg/minio/minio.go
package minio

import (
	"bytes"
	"context"
	"fmt"
	"time"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"github.com/skr1ms/SkyPostDelivery/drone-service/config"
)

type Client struct {
	client     *minio.Client
	bucketName string
}

func New(cfg *config.MinIO) (*Client, error) {
	minioClient, err := minio.New(cfg.Endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(cfg.RootUser, cfg.RootPassword, ""),
		Secure: cfg.UseSSL,
	})
	if err != nil {
		return nil, fmt.Errorf("MinioClient - New - New: %w", err)
	}

	return &Client{
		client:     minioClient,
		bucketName: cfg.BucketRecords,
	}, nil
}

func (c *Client) UploadFrame(ctx context.Context, droneID, deliveryID string, frameData []byte, frameNumber int) (string, error) {
	timestamp := time.Now().Format("20060102_150405")
	objectName := fmt.Sprintf("%s/%s/%s_frame_%d.jpg", droneID, deliveryID, timestamp, frameNumber)

	reader := bytes.NewReader(frameData)

	_, err := c.client.PutObject(ctx, c.bucketName, objectName, reader, int64(len(frameData)), minio.PutObjectOptions{
		ContentType: "image/jpeg",
	})
	if err != nil {
		return "", fmt.Errorf("MinioClient - UploadFrame - PutObject: %w", err)
	}

	return objectName, nil
}
